pkg/controller/database/postgres/postgres.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	ip := svc.Spec.ClusterIP
+	e.logger.Debug("postgres service", "ip", fmt.Sprintf("%+v", ip))
+
+	ps.SetConditions(runtimev1alpha1.Available())
+
+	return managed.ExternalObservation{ConnectionDetails: map[string][]byte{
+		runtimev1alpha1.ResourceCredentialsSecretEndpointKey: []byte(ip),
+	}, ResourceExists: true, ResourceUpToDate: true}, nil
+}
+
+func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
+	ps, ok := mgd.(*v1alpha1.Postgres)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
+	}
+	pvc, err := postgres.MakePVCPostgres(ps)
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errPVCCreateMsg)
+	}
+	e.logger.Debug("pvc make", "pvc", fmt.Sprintf("%+v", pvc))
+	if _, err := e.client.CreateOrUpdate(ctx, pvc); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errPVCCreateMsg)
+	}
+	// deploy credentials secret
+	password, err := e.client.ParseInputSecret(ctx, *ps)
+	if err != nil || password == "" {
+		password, err = e.client.GeneratePassword()
+		if err != nil {
+			return managed.ExternalCreation{}, errors.Wrap(err, errGeneratePasswordMsg)
+		}
+	}
+
+	// deploy deployment
+	if _, err := e.client.CreateOrUpdate(ctx, postgres.MakePostgresDeployment(ps, password)); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errDeployCreateMsg)
+	}
+	// deploy service
+	if _, err := e.client.CreateOrUpdate(ctx, postgres.MakeDefaultPostgresService(ps)); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errSVCCreateMsg)
+	}
+
+	return managed.ExternalCreation{
+		ConnectionDetails: map[string][]byte{
+			runtimev1alpha1.ResourceCredentialsSecretUserKey:     []byte(postgres.StringPtrToVal(ps.Spec.ForProvider.MasterUsername)),
+			runtimev1alpha1.ResourceCredentialsSecretPasswordKey: []byte(password),
+			runtimev1alpha1.ResourceCredentialsSecretPortKey:     []byte(strconv.Itoa(utils.IntValue(ps.Spec.ForProvider.Port))),
+			ResourceCredentialsSecretDatabaseKey:                 []byte(postgres.StringPtrToVal(ps.Spec.ForProvider.Database)),
+		},
+	}, nil
+}
+
+func (e *external) Update(ctx context.Context, mgd resource.Managed) (managed.ExternalUpdate, error) {
+	_, ok := mgd.(*v1alpha1.Postgres)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errUnexpectedObject)
+	}
+	return managed.ExternalUpdate{}, nil
+}
+
+func (e *external) Delete(ctx context.Context, mgd resource.Managed) error {
+	ps, ok := mgd.(*v1alpha1.Postgres)
+	if !ok {
+		return errors.New(errUnexpectedObject)
+	}
+	err := e.client.DeleteBucketService(ctx, ps)
+	if err != nil {
+		return errors.Wrap(err, errDelete)
+	}
+	err = e.client.DeleteBucketDeployment(ctx, ps)
+	if err != nil {
+		return errors.Wrap(err, errDelete)
+	}
+	err = e.client.DeleteBucketPVC(ctx, ps)
+	return errors.Wrap(err, errDelete)
+}
+
+func initializeDefaults(pg *v1alpha1.Postgres) bool {
+	updated := false
+	if pg.Namespace == "" {
```

Comment:
should you trim this for whitespace? just thinking of typos like accidentally adding a space " " ?

Reply:
That's fair, I added a trim now. I also added a note for why this step is important :)